dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+        CancellationToken? cancel = null)
+    {
+        context ??= new SKContext(new ContextVariables(), null!, null, log ?? NullLogger.Instance, cancel ?? CancellationToken.None);
+
+        if (this.Function is not null)
+        {
+            var result = await this.Function.InvokeAsync(context, settings, log, cancel);
+
+            if (result.ErrorOccurred)
+            {
+                result.Log.LogError(
+                    result.LastException,
+                    "Something went wrong in plan step {0}.{1}:'{2}'", this.SkillName, this.Name, context.LastErrorDescription);
+                return result;
+            }
+
+            context.Variables.Update(result.Result.ToString());
+        }
+        else
+        {
+            // loop through steps and execute until completion
+            while (this.HasNextStep)
+            {
+                var functionContext = context;
+                // Loop through State and add anything missing to functionContext
+
+                foreach (var item in this.State)
+                {
+                    if (!functionContext.Variables.ContainsKey(item.Key))
+                    {
+                        functionContext.Variables.Set(item.Key, item.Value);
+                    }
+                }
+
+                await this.InvokeNextStepAsync(functionContext);
+
+                context.Variables.Update(this.State.ToString());
+            }
+        }
+
+        return context;
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetDefaultSkillCollection(IReadOnlySkillCollection skills)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetDefaultSkillCollection(skills);
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetAIService(Func<ITextCompletion> serviceFactory)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetAIService(serviceFactory);
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetAIConfiguration(CompleteRequestSettings settings)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetAIConfiguration(settings);
+    }
+
+    #endregion ISKFunction implementation
+
+    /// <summary>
+    /// Invoke the next step of the plan
+    /// </summary>
+    /// <param name="context">Context to use</param>
+    /// <returns>The updated plan</returns>
+    /// <exception cref="KernelException">If an error occurs while running the plan</exception>
+    public async Task<Plan> InvokeNextStepAsync(SKContext context)
+    {
+        if (this.HasNextStep)
+        {
+            var step = this.Steps[this.NextStep];
+
+            context = await step.InvokeAsync(context);
+
+            if (context.ErrorOccurred)
+            {
+                throw new KernelException(KernelException.ErrorCodes.FunctionInvokeError,
```

Comment:
This is a planning exception. Kernel Exceptions should be thrown from SKFunction Invoke or Kernel itself.

Reply:
Not sure it is a planning exception... this is raised as a result of the kernel throwing a kernel exception.

Reply:
Gonna leave it for now

Reply:
Yeah, planning == creating the plan. This is execution of the plan by the kernel.